### ISSUE TITLE
Предложение по экстракодам.

### ISSUE THREAD

**Issue by mrp149** [besm6/mesm6#3]
Несколько моментов было бы интересно реализовать.

    Предлагается, добавить еще 32Ксл-Команд + 32Ксл-Данных (экстра сегменты) исключительно для поля реализации экстакоров, т.е. экстракод (в пользовательском режиме) - полный эквивалент обычной инструкции, но исполнение его отличается тем, что идет переключение выполнения в экстра сегмент по вектору соответствующего КОП-а. Микрокод только разбирает команду и выполняет всю адресную арифметику для удобства реализации экстакода. Принципы выполнения инструкций в экстракод такие-же, как и в режиме пользователя (полностью! со всеми вытекающими последствиями), внимание! реализация экстакода, может использовать экстракоды. Для чего это было-бы удобно? Например: было бы возможно реализовывать командные расширения, в зависимости от приложения, байтовые операции, криптография, специальные функции и ввод-вывод.
    Обработку прерываний возможно реализовать, на том же векторном принципе, но использовать вектора КОПов обычых команд. 046 векторов для коротко-адресных и 016 векторов для длинных или 48+4 векторов.
    Возврат из экстракода и прерывания возможно реализовать абсолютно идентично по спец. команде, (*36?)

**Comment by sergev** [comment]
1. Дополнительные сегменты предполагают изоляцию по данным от режима пользователя. Мне кажется, для приложений типа Arduino это не особо осмысленно. Я предполагаю реализовать экстракоды как прерывания с переходом на адреса, соответствующие коду команды. То есть 00046-00077 и 00020-00021, очень просто запомнить. Да, из экстракода можно будет вызывать другие экстракоды.

1. Для внешний прерываний будут адреса 00010-00017. При сбросе выполнение начинается с 00001.

1. Возврат из экстракодов и прерываний будет одинаковый: команда 32 (IJ, она же ВЫПР).

**Comment by sergev** [comment]
Есть идея ввести для режима прерываний/экстракодов отдельный набор регистров K[1]...K[15]. В режиме пользователя (после ВЫПР) возможны обращения только к M[1]...M[15]. При переключении в режим прерываний обращения идут к регистрам K[1]...K[15]. Из режима прерываний можно читать-писать регистры пользователя командами ITA/ATI, обращаясь к ним как к регистрам K[17]...K[31].

Преимущества:
1. Обработчики прерывания имеют отдельный стек, задаваемый регистром K[15].

1. Не надо упрятывать регистры в стеке. Ускоряется обработка прерываний. Достаточно упрятать сумматор и РМР.

1. Не нужны дополнительные регистры для сохранения информации о прерывании. Например, прерванный счётчик команд PC прячется в K[1], регистр режимов пользователя - в K[2], исполнительный адрес экстракода - в K[3].

**Comment by mrp149** [comment]
Да,  с памятью без приписки слазит к пользователю нужно будет делать  АSI регистр, т.е. доп. гимор, собственно это не сложно, конечно, но на бэсм так код  не писали.  Сегмент экстракода  хорош тем, что он изолирован и не видим, что создает возможность  вносить необходимые расширения для приложений непосредственно в ОС .  Ладно, оставим  сегменты для будущего расширения.  
Отдельный набор регистров для прерываний, был заделан на микро-бэсм, см секцию 4.8 и посдсекции 4.8.1-4.8.2 про то, как были реализованы экстракоды. Сделать надо бы проще, чем на микро-бэсм. Возможно надо будет расширять РР, если будет доп. набор, то придется ставить признаки в РР какой набор использовать на возврате, как-бы с РР не начал расти ком!

**Comment by x86128** [comment]
>     1. Да, из экстракода можно будет вызывать другие экстракоды.

>     3. Не нужны дополнительные регистры для сохранения информации о прерывании. Например, прерванный счётчик команд PC прячется в K[1], регистр режимов пользователя - в K[2], исполнительный адрес экстракода - в K[3].

Что делать с текущим К1 при вызове вложенного экстракода?

Можно сделать Экстракоды так чтобы они всегда выталкивали полный текущий PC_next по адресу [K15--], а команда ВЫПР возвращалась на [++K15]. Программное выталкивание К2 прописать в "соглашении о программных вызовах".

**Comment by x86128** [comment]
> * Я предполагаю реализовать экстракоды как прерывания с переходом на адреса, соответствующие коду команды. То есть 00046-00077 и 00020-00021
> 
> * Для внешний прерываний будут адреса 00010-00017.

Поскольку в векторе лежит пара команд, это позволит сделать и "динамические" вектора прерываний:

`wtc dynvec1, uj 0` где содержимое ячейки **dynvec1** меняется по ходу исполнения программы.

**Comment by mrp149** [comment]
Вложенные экстракоды - отдельный случай... 
Идеально, было бы иметь регистр с базовым адресом векторов, т.е. конкретный вектор - сдвиг от базового адреса. На сбросе значение базового адреса = 0
